bigquery_etl/backfill/shredder_mitigation.py

```diff
@@ -0,0 +1,196 @@
+    """Representation of a column in a query, with relevant details for shredder mitigation."""
+
+    def __init__(self, name, data_type, column_type, status):
+        """Initialize class with required attributes."""
+        self.name = name
+        self.data_type = data_type
+        self.column_type = column_type
+        self.status = status
+
+    def __eq__(self, other):
+        """Return attributes only if the referenced object is of type Column."""
+        if not isinstance(other, Column):
+            return NotImplemented
+        return (
+            self.name == other.name
+            and self.data_type == other.data_type
+            and self.column_type == other.column_type
+            and self.status == other.status
+        )
+
+    def __repr__(self):
+        """Return a string representation of the object."""
+        return f"Column(name={self.name}, data_type={self.data_type}, column_type={self.column_type}, status={self.status})"
+
+
+def get_bigquery_type(value) -> DataTypeGroup:
+    """Return the datatype of a value, grouping similar types."""
+    date_formats = [
+        ("%Y-%m-%d", date),
+        ("%Y-%m-%d %H:%M:%S", datetime),
+        ("%Y-%m-%dT%H:%M:%S", datetime),
+        ("%Y-%m-%dT%H:%M:%SZ", datetime),
+        ("%Y-%m-%d %H:%M:%S UTC", datetime),
+        ("%H:%M:%S", time),
+    ]
+    for format, dtype in date_formats:
+        try:
+            if dtype == time:
+                parsed_to_time = datetime.strptime(value, format).time()
+                if isinstance(parsed_to_time, time):
+                    return DataTypeGroup.DATE
+            else:
+                parsed_to_date = datetime.strptime(value, format)
+                if isinstance(parsed_to_date, dtype):
+                    return DataTypeGroup.DATE
+        except (ValueError, TypeError):
+            continue
+    if isinstance(value, time):
+        return DataTypeGroup.DATE
+    if isinstance(value, date):
+        return DataTypeGroup.DATE
+    elif isinstance(value, bool):
+        return DataTypeGroup.BOOLEAN
+    if isinstance(value, int):
+        return DataTypeGroup.NUMERIC
+    elif isinstance(value, float):
+        return DataTypeGroup.FLOAT
+    elif isinstance(value, str) or isinstance(value, bytes):
+        return DataTypeGroup.STRING
+    elif isinstance(value, NoneType):
+        return DataTypeGroup.UNDETERMINED
+    else:
+        raise ValueError(f"Unsupported data type: {type(value)}")
+
+
+def classify_columns(
+    new_row: dict, existing_columns: list, new_columns: list
+) -> tuple[list[Column], list[Column], list[Column], list[Column], list[Column]]:
+    """Compare the new row with the existing columns and return the list of common, added and removed columns by type."""
+    common_dimensions = []
+    added_dimensions = []
+    removed_dimensions = []
+    metrics = []
+    undefined = []
+
+    for key in existing_columns:
+        if key in existing_columns and key not in new_columns:
```

Comment:
in line 127 you have `existing_columns` and in line 128 as well ... one of those should probably be the `new_row` instead?

Reply:
```suggestion
        if key not in new_columns:
```